Assets/MixedRealityToolkit-SDK/Features/Diagnostics/DiagnosticsHandler.cs

```diff
@@ -0,0 +1,124 @@
+
+namespace Microsoft.MixedReality.Toolkit.SDK.DiagnosticsSystem
+{
+    public class DiagnosticsHandler : MonoBehaviour, IMixedRealityDiagnosticsHandler
+    {
+        private bool showCpu;
+        private bool ShowCpu
+        {
+            get { return showCpu; }
+            set
+            {
+                if (showCpu != value)
+                {
+                    showCpu = value;
+                    if (!showCpu)
+                    {
+                        cpuUseTracker.Reset();
+                    }
+                }
+            }
+        }
+
+        private bool ShowFps { get; set; }
+        private bool ShowMemory { get; set; }
+
+        private string displayText;
+        private int numberOfLines = 0;
+        private bool isShowingInformation;
+
+        private CpuUseTracker cpuUseTracker = new CpuUseTracker();
+        private MemoryUseTracker memoryUseTracker = new MemoryUseTracker();
+        private FpsUseTracker fpsUseTracker = new FpsUseTracker();
+
+        public void OnDiagnosticSettingsChanged(DiagnosticsEventData eventData)
+        {
+            this.ShowCpu = eventData.ShowCpu;
+            this.ShowMemory = eventData.ShowMemory;
+            this.ShowFps = eventData.ShowFps;
+            this.enabled = eventData.Visible;
+        }
+
+        void UpdateIsShowingInformation()
+        {
+            isShowingInformation = ShowCpu ||
+                                   ShowFps ||
+                                   ShowMemory;
+        }
+        
+        void Update()
+        {
+            UpdateIsShowingInformation();
+
+            if (!isShowingInformation)
+            {
+                return;
+            }
+
+            StringBuilder sb = new StringBuilder();
+            numberOfLines = 0;
+
+            if (ShowFps)
+            {
+                var timeInSeconds = fpsUseTracker.GetFpsInSeconds();
+                sb.AppendLine($"Fps: {Math.Round(1.0f / timeInSeconds, 2)}");
+                sb.AppendLine($"Frame Time: {Math.Round(timeInSeconds * 1000, 2)} ms");
+                numberOfLines++;
+            }
+
+            if (ShowCpu)
+            {
+                var reading = cpuUseTracker.GetReadingInMs();
+                sb.AppendLine($"CPU Time: {reading} ms");
+                numberOfLines++;
+            }
+
+            if (ShowMemory)
+            {
+                var reading = memoryUseTracker.GetReading();
+                sb.AppendLine($"Memory: {Math.Round(BytesToMB(reading.GCMemoryInBytes), 2)} MB");
+                numberOfLines++;
+            }
+
+            if (numberOfLines > 0)
+            {
+                displayText = sb.ToString();
+            }
+            else
+            {
+                displayText = null;
+            }
+        }
+
+        void OnGUI()
+        {
+            if (!isShowingInformation || displayText == null)
+            {
+                return;
+            }
+
+            var style = new GUIStyle();
```

Comment:
let's get rid of all these new keywords in our update loops.

We can cache and update a lot of this stuff.

Reply:
we can create it, then check if it's not null after.

Reply:
Just did a quick and dirty test to see how much stack allocation actually costs, and newing a struct was twice as expensive with int.MaxValue iterations (both were still really quick though). I'll move this to a field and remove the new here. Good catch

Reply:
We also need to be aware that imgui is depreciated and it's most likely getting taken out of the core Unity Engine soon.

Reply:
I don't see where it's marked deprecated, but I believe we should open another issue on that if it is. I was looking at https://docs.unity3d.com/2018.3/Documentation/Manual/GUIScriptingGuide.html for information.